logstash-core/src/main/java/org/logstash/config/ir/CompiledPipeline.java

```diff
@@ -0,0 +1,388 @@
+     * a {code filter} or and {code if} statement).
+     * @param parents Nodes from the last already compiled level
+     * @param start Vertex to compile children for
+     * @return Datasets originating from given {@link Vertex}
+     */
+    private Collection<Dataset> flatten(final Collection<Dataset> parents, final Vertex start) {
+        final Collection<Vertex> endings = start.incomingVertices()
+            .filter(v -> isFilter(v) || isOutput(v) || v instanceof IfVertex)
+            .collect(Collectors.toList());
+        return endings.isEmpty() ? parents : flattenChildren(parents, start, endings);
+    }
+
+    /**
+     * Compiles all child vertices for a given vertex.
+     * @param parents Parent datasets from previous stage
+     * @param start Start Vertex that got expanded
+     * @param children Children of {@code start} that are either {@code if} or {@code filter} in
+     * type
+     * @return Datasets compiled from vertex children
+     */
+    private Collection<Dataset> flattenChildren(final Collection<Dataset> parents,
+        final Vertex start, final Collection<Vertex> children) {
+        return children.stream().map(
+            child -> {
+                final Collection<Dataset> newparents = flatten(parents, child);
+                if (isFilter(child)) {
+                    return filterDataset(child.getId(), newparents);
+                } else if (isOutput(child)) {
+                    return outputDataset(child.getId(), newparents);
+                } else {
+                    // We know that it's an if vertex since the the input children are either 
+                    // output, filter or if in type.
+                    final IfVertex ifvert = (IfVertex) child;
+                    final EventCondition iff = buildCondition(ifvert);
+                    final String index = ifvert.getId();
+                    // It is important that we double check that we are actually dealing with the
+                    // positive/left branch of the if condition
+                    if (ifvert.getOutgoingBooleanEdgesByType(true).stream()
+                        .anyMatch(edge -> Objects.equals(edge.getTo(), start))) {
+                        return splitLeft(newparents, iff, index);
+                    } else {
+                        return splitRight(newparents, iff, index);
+                    }
+                }
+            }).collect(Collectors.toList());
+    }
+
+    /**
+     * Build a {@link Dataset} representing the {@link JrubyEventExtLibrary.RubyEvent}s after
+     * the application of the given filter.
+     * @param vertex Vertex Id of the filter to create this {@link Dataset} for
+     * @param parents All the parent nodes that go into this filter
+     * @return Filter {@link Dataset}
+     */
+    private Dataset filterDataset(final String vertex, final Collection<Dataset> parents) {
+        final Dataset filter;
+        if (plugins.containsKey(vertex)) {
```

Comment:
Also a good place for `Map.computeIfAbsent`

Reply:
jup :) fixing